boot/boot.go

```diff
@@ -0,0 +1,171 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2019 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package boot
+
+import (
+	"errors"
+	"fmt"
+	"path/filepath"
+	"strings"
+
+	"github.com/snapcore/snapd/bootloader"
+	"github.com/snapcore/snapd/logger"
+	"github.com/snapcore/snapd/snap"
+)
+
+// A Device Boot Set is a collection of snaps that are involved in a
+// device's boot process.
+type DeviceBootSet interface {
+	// RemoveKernelAssets removes the unpacked kernel/initrd for the given
+	// kernel snap.
+	RemoveKernelAssets() error
+	// ExtractKernelAssets extracts kernel/initrd/dtb data from the given
+	// kernel snap, if required, to a versioned bootloader directory so
+	// that the bootloader can use it.
+	ExtractKernelAssets(snap.Container) error
+	// SetNextBoot will schedule the given OS or base or kernel snap to be
+	// used in the next boot. For base snaps it up to the caller to select
+	// the right bootable base (from the model assertion).
+	SetNextBoot() error
+	// ChangeRequiresReboot returns whether a reboot is required to switch
+	// to the given OS, base or kernel snap.
+	ChangeRequiresReboot() bool
+}
+
+// Lookup figures out what the right boot set is for the given
+// arguments, and returns it.
+func Lookup(s snap.PlaceInfo, t snap.Type, onClassic bool) (DeviceBootSet, error) {
```

Comment:
the error is only used to express that the is unexpected.  Let's make this:

`Lookup(...) (bootp BootParticipant, applicable bool)`

this will return `nil, false` for not relevant types.  And a Kernel if the snap is a kernel.

Then in the caller we can two kind of checks:

```
if applicable { ... }
```

```
if krnl, ok := bootp.(boot.Kernel); ok { ... }
```